src/librustc_resolve/build_reduced_graph.rs

```diff
@@ -700,13 +700,12 @@ impl<'a, 'b> BuildReducedGraphVisitor<'a, 'b> {
             }
 
             // These items live in the type namespace.
-            ItemKind::TyAlias(..) => {
-                let res = Res::Def(DefKind::TyAlias, self.r.definitions.local_def_id(item.id));
-                self.r.define(parent, ident, TypeNS, (res, vis, sp, expansion));
-            }
-
-            ItemKind::OpaqueTy(_, _) => {
-                let res = Res::Def(DefKind::OpaqueTy, self.r.definitions.local_def_id(item.id));
+            ItemKind::TyAlias(ref ty, _) => {
+                let def_kind = match ty.kind.opaque_top_hack() {
+                    None => DefKind::TyAlias,
+                    Some(_) => DefKind::OpaqueTy,
```

Comment:
So, I noticed today that `DefKind::OpaqueTy` and `DefKind::AssocOpaqueTy` exist despite the `existential type` syntax being removed.

Is it possible to remove them after this PR or it will require reworking lowering as well?

Reply:
I'm not sure, we might need something like:
```rust
type Foo = (impl Bar, impl Baz); // In AST
```
=> (HIR)
```rust
opaque type _0: Bar; // DefKind::OpaqueTy
opaque type _1: Baz; // DefKind::OpaqueTy
type Foo = ( // DefKind::TyAlias
     _0, // TyKind::Def(..)
     _1 // TyKind::Def(..)
);
```

Reply:
From my point of view: as long we still have something like `opaque type` in HIR still, that's good enough for my upcoming PR. I'd rather not remove that for a while yet.

Reply:
We won't get rid of opaque types in HIR (they will always just arise from type aliases).

Reply:
Yeah, that's what I thought. :-)